subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java

```diff
@@ -42,14 +43,37 @@ public CompositeAwareTaskSelector(GradleInternal gradle, BuildStateRegistry buil
     @Override
     public Spec<Task> getFilter(String path) {
         Path taskPath = Path.path(path);
-        BuildState build = findIncludedBuild(taskPath);
-        if (build != null) {
-            return getSelector(build).getFilter(taskPath.removeFirstSegments(1).toString());
+        if (taskPath.isAbsolute()) {
+            if (gradle.isRootBuild()) {
+                if (pathReferencesAnyIncludedBuilds(taskPath)) {
+                    return Specs.satisfyAll();
+                } else {
+                    return getUnqualifiedBuildSelector().getFilter(path);
+                }
+            } else {
+                if (pathReferencesThisIncludedBuild(taskPath)) {
+                    return getUnqualifiedBuildSelector().getFilter(taskPath.removeFirstSegments(1).toString());
+                } else {
+                    return Specs.satisfyAll();
+                }
+            }
         } else {
-            return getUnqualifiedBuildSelector().getFilter(path);
+            if (gradle.isRootBuild()) {
+                return getUnqualifiedBuildSelector().getFilter(path);
+            } else {
+                return Specs.satisfyAll();
+            }
         }
     }
 
+    private boolean pathReferencesAnyIncludedBuilds(Path taskPath) {
+        return findIncludedBuild(taskPath) != null;
+    }
+
+    private boolean pathReferencesThisIncludedBuild(Path taskPath) {
+        return taskPath.segmentCount() > 1 && taskPath.segment(0).equals(gradle.getRootProject().getName());
+    }
```

Comment:
This read a little weird to me. I think this basically does the same thing, but uses the existing `findIncludedBuild` method.  WDYT?
```
    public Spec<Task> getFilter(String path) {
        Path taskPath = Path.path(path);
        if (taskPath.isAbsolute()) {
            BuildState build = findIncludedBuild(taskPath);
            // Exclusion was for an included build, use it
            if (build != null) {
                return getSelector(build).getFilter(taskPath.removeFirstSegments(1).toString());
            }
        }
        // Exclusion didn't match an included build, so it might be a subproject of the root build or a relative path
        if (gradle.isRootBuild()) {
            return getUnqualifiedBuildSelector().getFilter(path);
        } else {
            // Included build ignores this exclusion since it doesn't apply directly to it
            return Specs.satisfyAll();
        }
    }
```

Reply:
I've tried several variants on what could be the most readable, but none of them was satisfying enough. I think I'll just go with your solution.